# Customer Retention Analysis

## Libraries

In [2]:
import numpy as np
import pandas as pd
import matplotlib as mpl
import plotly.express as px

from nb_pyscript.data import data_overview
from nb_pyscript.visualize import pie_chart, bar_chart

Basic settings of dataframe and plots

In [3]:
# Pandas dataframes show all rows and columns without skiping
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

# Matplotlib setting of figsize and dpi
mpl.rcParams['figure.figsize'] = [16, 5]
mpl.rcParams['figure.dpi'] = 200

# Colors
red = "#fa5252"
green = "#40c057"

%matplotlib inline

## Dataset

In [4]:
data = pd.read_excel('../data/telco-customer-churn/telco_customer_churn.xlsx')

## Data Exploration

In [5]:
df = data.copy()

In [6]:
df.head()

CustomerID  Count        Country       State         City  Zip Code  \
0  3668-QPYBK      1  United States  California  Los Angeles     90003   
1  9237-HQITU      1  United States  California  Los Angeles     90005   
2  9305-CDSKC      1  United States  California  Los Angeles     90006   
3  7892-POOKP      1  United States  California  Los Angeles     90010   
4  0280-XJGEX      1  United States  California  Los Angeles     90015   

                 Lat Long   Latitude   Longitude  Gender Senior Citizen  \
0  33.964131, -118.272783  33.964131 -118.272783    Male             No   
1   34.059281, -118.30742  34.059281 -118.307420  Female             No   
2  34.048013, -118.293953  34.048013 -118.293953  Female             No   
3  34.062125, -118.315709  34.062125 -118.315709  Female             No   
4  34.039224, -118.266293  34.039224 -118.266293    Male             No   

  Partner Dependents  Tenure Months Phone Service Multiple Lines  \
0      No         No              2           Yes             No   
1      No        Yes              2           Yes             No   
2      No        Yes              8           Yes            Yes   
3     Yes        Yes             28           Yes            Yes   
4      No        Yes             49           Yes            Yes   

  Internet Service Online Security Online Backup Device Protection  \
0              DSL             Yes           Yes                No   
1      Fiber optic              No            No                No   
2      Fiber optic              No            No               Yes   
3      Fiber optic              No            No               Yes   
4      Fiber optic              No           Yes               Yes   

  Tech Support Streaming TV Streaming Movies        Contract  \
0           No           No               No  Month-to-month   
1           No           No               No  Month-to-month   
2           No          Yes              Yes  Month-to-month   
3          Yes          Yes              Yes  Month-to-month   
4           No          Yes              Yes  Month-to-month   

  Paperless Billing             Payment Method  Monthly Charges Total Charges  \
0               Yes               Mailed check            53.85        108.15   
1               Yes           Electronic check            70.70        151.65   
2               Yes           Electronic check            99.65         820.5   
3               Yes           Electronic check           104.80       3046.05   
4               Yes  Bank transfer (automatic)           103.70        5036.3   

  Churn Label  Churn Value  Churn Score  CLTV                   Churn Reason  
0         Yes            1           86  3239   Competitor made better offer  
1         Yes            1           67  2701                          Moved  
2         Yes            1           86  5372                          Moved  
3         Yes            1           84  5003                          Moved  
4         Yes            1           89  5340  Competitor had better devices

In [7]:
data_overview(df, "Overview of telco customer data:")

Overview of telco customer data:


Entries in the dataset: 7043
Features in the dataset: 33


Faatures of the dataset:
 ['CustomerID', 'Count', 'Country', 'State', 'City', 'Zip Code', 'Lat Long', 'Latitude', 'Longitude', 'Gender', 'Senior Citizen', 'Partner', 'Dependents', 'Tenure Months', 'Phone Service', 'Multiple Lines', 'Internet Service', 'Online Security', 'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV', 'Streaming Movies', 'Contract', 'Paperless Billing', 'Payment Method', 'Monthly Charges', 'Total Charges', 'Churn Label', 'Churn Value', 'Churn Score', 'CLTV', 'Churn Reason']


Missing Values:
CustomerID                    : 0
Count                         : 0
Country                       : 0
State                         : 0
City                          : 0
Zip Code                      : 0
Lat Long                      : 0
Latitude                      : 0
Longitude                     : 0
Gender                        : 0
Senior Citizen                :

**About the dataset:**

- Dataset contains 7043 entries and 33 features.

- Features of the dataset are-
    - `Customer ID`: ID of customers each are unique
    - `Count`: Frequency of the customer in the table
    - `Country`: Country of the customer
    - `State`: State where customer lives
    - `City`: City where customer resides(1129 unique city names)
    - `Zip Code`: Zip code of the place
    - `Lat Long`: Latitude and Longitude of the place
    - `Latitude`: Latitude of the place
    - `Longitude`: Longitude of the place
    - `Gender`: Gender of the customer (Male or Female)
    - `Senior Citizen`: Indicates whether the client is an older person (0, 1).
    - `Partner`: Indicates whether or not the client is partnered (Yes, No).
    - `Dependents`: Indicates whether the client is supported by others (Yes, No).
    - `Tenure Months`: The length of time a customer has been a customer of the business (multiple different number values).
    - `Phone Service`: If the client has a phone service, it is indicated by the words “Yes” or “No”.
    - `Multiple Line`: Whether the customer has more than one line (no phone service, no service, yes service).
    - `Internet Service`: Whether the client has a subscription to the company’s Internet service (DSL, Fiber optic, or No).
    - `Online Security`: Indicates if the client has access to online security (Internet service available, No, Yes).
    - `Online Backup`: Indicates whether or not the client has an online backup (Internet service unavailable, No, Yes).
    - `Device Protection`: Indicates whether the client has device protection (Internet service not available, Not Available, Yes).
    - `Tech Support`: Whether the customer has access to tech help (no internet service, no, yes).
    - `Streaming TV`: Whether the customer has access (no internet service, no, yes).
    - `Streaming Movies`: Streaming movies: Indicates whether or not the client offers or has access streaming movies(no internet service, no, yes).
    - `Contract`: The type of existing contract for the customer (Month-to-Month, One-Year, Two-Year)
    - `Paperless Billing`: Whether the client uses paperless billing (Yes, No)
    - `Payment Method`: The chosen payment method by the consumer (credit card, bank transfer, electronic check, paper check)
    - `Monthly Charges`: The monthly charge made to the consumer (various numeric quantities)
    - `Total Charges`: (many different numeric values): The total amount charged to the consumer
    - `Churn Label`: Whether the customer decided to leave the service or not ('yes' or 'no').
    - `Churn Value`: 0 or 1 depending on the churn value as 'no' or 'yes'.
    - `CLTV`: Customer Lifetime Value if they stayed in the company as a customer.
    - `Churn Reason`: Churn Reason: Reason for the customer to choose to leave.

- Dataset also contains some numerical features like-
    - `Tenure Months`, `Monthly Charges` and `Total Charges`

- Apart from the numerical features mentioned here, there some other features which are redundant like-
    - `Count`, `Country`, `State`

- Also we will not use `Zip Code`, `Lat Long`, `Latitude` and `Longitude` as they won't provide much information becase there are a lot of values in them if we treat it as categorical data.

- Dataset contains target label `Churn Label` and `Churn Value`. As both of them are representing same thing, I am removing `Churn Label`.

- In this project, I will totally focus on churn prediction and analysis and avoid any complications so I will also remove `Churn Reason` and `CLTV`.

- The features can be categorized into 3 groups-
    - *Customer Demographic*: `City`, `Gender`, `Senior Citizen`, `Partner` and `Dependent`.
    - *Customer Account*: `Tenure Months`, `Contract`, `Paperless Billing`, `Payment Method`, `Monthly Charges` and `Total Charges`.
    - *Services*: `Phone Service`, `Multiple Line`, `Internet Service`, `Online Security`, `Online Backup`, `Device Protection`, `Tech Support`, `Streaming TV`, `Streaming Movies`.

In [8]:
# Drop the redundant columns from the dataset
columns = ['Count', 'Country', 'State', 'Zip Code', 'Lat Long', 'Latitude', 'Longitude', 'Churn Reason', 'Churn Score', 'Churn Value', 'CLTV']

non_redundant_df = df.drop(columns=columns)
non_redundant_df.columns

Index(['CustomerID', 'City', 'Gender', 'Senior Citizen', 'Partner',
       'Dependents', 'Tenure Months', 'Phone Service', 'Multiple Lines',
       'Internet Service', 'Online Security', 'Online Backup',
       'Device Protection', 'Tech Support', 'Streaming TV', 'Streaming Movies',
       'Contract', 'Paperless Billing', 'Payment Method', 'Monthly Charges',
       'Total Charges', 'Churn Label'],
      dtype='object')

In [9]:
df = non_redundant_df.copy()

In [10]:
# Pie chart to see the distribution of churn and not churn customers
fig = pie_chart(df, feature="Churn Label", colors=[green, red], hole=0.3, title="Distribution of Churn and Not Churn Customers", annotation_text="")

fig.show()

From the above pie chart,
- Churned Customers: `26.5%`
- Not Churned Customers: `73.5%`

In [34]:
# Bar chart to show the distribution of the categorical features in the dataset
fig = bar_chart(df, x="Gender", groupby_feature="Churn Label", colors=[green, red], title="Distribution of Gender for each Churn Label", annotation_text=True)
fig.show()